### DISCUSSION TITLE
bug: inference error when creating a procedure with middleware: The inferred type of 'publicProcedure' cannot be named without a reference to '.pnpm/...'. This is likely not portable. A type annotation is necessary.

### DISCUSSION THREAD

**Discussion by MiniSuperDev** [trpc/trpc#4906] [Q&A]
### Provide environment information

```
  System:
    OS: Windows 10 10
    CPU: (4) x64 Intel(R) Core(TM) i7
    Memory: 6.99 GB / 15.89 GB
  Binaries:
    Node: 18.17.1 - C:\Program Files\nodejs\node.EXE    
    Yarn: 1.22.19 - ~\AppData\Roaming\npm\yarn.CMD      
    npm: 9.6.7 - C:\Program Files\nodejs\npm.CMD        
    pnpm: 8.7.4 - ~\AppData\Local\pnpm\pnpm.EXE
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (117.0.2045.60)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @trpc/server: ^10.38.5 => 10.38.5
    typescript: ^5.2.2 => 5.2.2
```

### Describe the bug

Using turborepo with pnpm I have this error when I want to create a middleware.

![image](https://github.com/trpc/trpc/assets/105388540/4b01a284-80ce-472f-b9c9-3c25a317c253)

```sh
The inferred type of 'publicProcedure' cannot be named without a reference to '.pnpm/@types+qs@6.9.8/node_modules/@types/qs'. This is likely not portable. A type annotation is necessary.ts(2742)
```

Thanks

### Link to reproduction

https://github.com/MiniSuperDev/turbo-trpc-bug/tree/12c469463c447299082b07c494cd0319f5016589

### To reproduce

I created a repository isolating the error, so clone the repo and check this line:
https://github.com/MiniSuperDev/turbo-trpc-bug/blob/12c469463c447299082b07c494cd0319f5016589/apps/express-api/src/trpc.ts#L15

I notice that I can solve it by remove the types from my tsconfig, but I don't want to do this
https://github.com/MiniSuperDev/turbo-trpc-bug/blob/12c469463c447299082b07c494cd0319f5016589/apps/express-api/tsconfig.json#L19
For example here remove `"vitest/globals"`

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4905">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4905/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/4905/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
This works:

```ts
import { inferAsyncReturnType, initTRPC } from "@trpc/server";
import type * as trpcExpress from "@trpc/server/adapters/express";

interface ContextOptions
  extends Partial<trpcExpress.CreateExpressContextOptions> {}

export async function createTRPCContext(opts: ContextOptions) {
  return opts;
}

export type Context = inferAsyncReturnType<typeof createTRPCContext>;

export const t = initTRPC.context<Context>().create();

export const appRouter = t.router({});

export const publicProcedure = t.procedure.use(async (opts) => {
  return await opts.next();
});

```

Not a bug AFAICT

**Reply by KATT** [reply]
Follow https://trpc.io/docs/server/context#inner-and-outer-context

**Reply by MiniSuperDev** [reply]
@KATT  Hi, the bug is when you use types in your TS config, like "vitest/globals", like in the repo.

https://github.com/MiniSuperDev/turbo-trpc-bug/blob/12c469463c447299082b07c494cd0319f5016589/apps/express-api/tsconfig.json#L19

Can you check please :)

**Reply by KATT** [reply]
I don't have more time to debug your project unfortunately, but I don't see how this is a problem with trpc. We don't import anything except in the adapters where we import express etc so if you get this type of error it's usually fixed by exporting something from the lib.
